### PR TITLE
Add the 'bufferSize' option to GridLayer to increase tiles kept (#4039)

### PR DESCRIPTION
**Pull request by IvanSanchez**
Basically reuses some of the ideas from #3551 and the square-by-default tileSize option.

### CODE REVIEW

**Comment by IvanSanchez** [comment]
Note to self: add `bufferSize: 0` to all unit tests involving tile counts, and fix #4613 in order to check locally.

**Comment by nathancahill** [comment]
Awesome. Really excited to see this.

**Comment by IvanSanchez** [comment]
@nathancahill Feel free to check out the branch and test ;-)

File: src/layer/tile/GridLayer.js [hunk 1]
```diff
@@ -256,6 +264,11 @@ L.GridLayer = L.Layer.extend({
 		return s instanceof L.Point ? s : new L.Point(s, s);
 	},
 
+	_getBufferSize: function () {
+		var s = this.options.bufferSize;
+		return s instanceof L.Point ? s : new L.Point(s, s);
```
Comment (mourner) [inline src/layer/tile/GridLayer.js]: I would rather initialize `_bufferSize` in constructor.

**Comment by mourner** [comment]
> Note to self: add bufferSize: 0 to all unit tests involving tile counts

Why add this if it's 0 by default anyway?

File: src/layer/tile/GridLayer.js (hunk 1 above)
Comment (IvanSanchez) [inline src/layer/tile/GridLayer.js]: Shall we do the same with `getTileSize` then?

**Comment by IvanSanchez** [comment]
> Why add this if it's 0 by default anyway?

I added a default of 256px of buffer, given the complaints from #4039.

**Comment by hyperknot** [comment]
The problem with this approach is that this loads the hidden tiles even if the user has no intention of viewing them. For example, on tiny 600x400 map embed (cca. 6 tiles) we are loading 28 tiles now. That can eat through a lot of tile credits quite quickly.

A better solution would be to tweak the tile _pruning_ code, such that it keeps _buffer_ tiles in memory instead of cleaning + reloading them later.

**Comment by pke** [comment]
@hyperknot yes the pruning code is way to aggressive imho. Panning back and forth a map view should not prune the borders of the map. Also for zooming it should keep more tiles in memory to make the experience seamless and not riddled with white tiles.

**Comment by IvanSanchez** [comment]
Superseded by #4650

**Pull request closed without merging**